### PR TITLE
FEATURE: Chat channel preview card

### PR DESCRIPTION
**Pull request by martin-brennan**
This commit adds a preview card for a channel that the current user
is previewing when they don't have a membership for the channel.
It shows the channel title + badge based on the category, and allows
the user to join the channel as well as return to the channel browse
list.

**Desktop - With Description**

**Desktop - No Description**

**Mobile - With Description**

**Mobile - No Description**

### CODE REVIEW

File: assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<div class="chat-channel-preview-card">
+  {{chat-channel-title channel=chatChannel}}
+  <p class="chat-channel-preview-card__description">{{or chatChannel.description "-"}}</p>
+  {{d-button
+    action=(action "onJoinChannel")
+    label="chat.channel_settings.join_channel"
+    class="chat-channel-preview-card__join-channel-button btn-primary"
+  }}
+  <p class="chat-channel-preview-card__browse-all">
+    <a href="/chat/browse">{{i18n "chat.browse_all_channels"}}</a>
```
Comment (jjaffeux) [inline assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs:10]: Maybe use a link-to here

File: assets/stylesheets/common/chat-channel-preview-card.scss [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    border-radius: 0.25rem;
+    line-height: normal;
+    box-sizing: border-box;
+    padding: 0.5em 0.65em;
+    font-weight: normal;
+    cursor: pointer;
+  }
+
+  &__browse-all {
+    font-size: .875rem;
```
Comment (jjaffeux) [inline assets/stylesheets/common/chat-channel-preview-card.scss:30]: Cant we use a var here?

File: assets/stylesheets/common/chat-channel-preview-card.scss [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  &__description {
+    color: var(--primary-600);
+  }
+
+  .chat-channel-title__name {
+    font-size: 1.25rem;
+  }
+
+  &__join-channel-button {
+    font-size: 1.25rem;
```
Comment (jjaffeux) [inline assets/stylesheets/common/chat-channel-preview-card.scss:19]: same cant we use a var?

File: assets/stylesheets/common/chat-channel-preview-card.scss [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+
+  &__description {
+    color: var(--primary-600);
+  }
+
+  .chat-channel-title__name {
+    font-size: 1.25rem;
```
Comment (jjaffeux) [inline assets/stylesheets/common/chat-channel-preview-card.scss:15]: var :p

File: assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<div class="chat-channel-preview-card">
+  {{chat-channel-title channel=chatChannel}}
+  <p class="chat-channel-preview-card__description">{{or chatChannel.description "-"}}</p>
+  {{d-button
+    action=(action "onJoinChannel")
+    label="chat.channel_settings.join_channel"
+    class="chat-channel-preview-card__join-channel-button btn-primary"
```
Comment (jjaffeux) [inline assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs:7]: we can go with -btn instead of -button

File: assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<div class="chat-channel-preview-card">
+  {{chat-channel-title channel=chatChannel}}
+  <p class="chat-channel-preview-card__description">{{or chatChannel.description "-"}}</p>
```
Comment (jjaffeux) [inline assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs:3]: do we want to show it at all if empty?

**Review by jjaffeux — COMMENTED** [review]
Sorry sa after it was a draft

**Comment by jjaffeux** [comment]
@martin-brennan as you will have to write tests for this, please have a look to what I did here with the help of @cvx: https://github.com/discourse/discourse-chat/commit/458f25368bc2d80c385ce010b958a6ae4c56fe17#diff-d2d0cde095125acf48a06ed57196e54d70385e056b753047c92eac44c112b4e6R112

We can finally use new latest ember test syntax and I would want us to use it for chat as much as possible.

File: assets/javascripts/discourse/components/chat-channel-about-view.js [hunk 7]
```diff
@@ -19,8 +19,13 @@ export default class ChatChannelAboutView extends Component {
 
     return ChatApi.followChatChannel(this.channel.id)
       .then((membership) => {
-        this.channel.set("following", true);
-        this.channel.set("memberships_count", membership.user_count);
+        this.channel.setProperties({
```
Comment (martin-brennan) [inline assets/javascripts/discourse/components/chat-channel-about-view.js:22]: This is now done in a few places in the app, maybe we can move it into the `followChatChannel` function, passing the whole channel object instead of just the ID?

**Comment by martin-brennan** [comment]
@jjaffeux can you please take a look at this one now that it's ready?

**Comment by martin-brennan** [comment]
Only test failures are to do with quoting, not sure why:

```
Error: QUnit Test Failure: Global error: Uncaught Error: Assertion Failed: expected container not to be destroyed at http://localhost:7357/assets/vendor.js, line 38378
Testem finished with non-zero exit code. Tests failed.
 While executing test: Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in
not ok 267 Chrome 103.0 - [undefined ms] - Global error: Uncaught Error: Assertion Failed: expected container not to be destroyed at http://localhost:7357/assets/vendor.js, line 38378
 While executing test: Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in
    ---
        browser log: |
            {"type":"error","text":"Uncaught Error: Assertion Failed: expected container not to be destroyed at http://localhost:7[357](https://github.com/discourse/discourse-chat/runs/7275262839?check_suite_focus=true#step:23:358)/assets/vendor.js, line 38378\n","testContext":{"id":6,"name":"Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in","items":[],"state":"complete"}}
    ...
Error: QUnit Test Failure: Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in
not ok 268 Chrome 103.0 - [562 ms] - Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in
    ---
        actual: >
            /chat/channel/7/Uncategorized
        expected: >
            /c/uncategorized/1
        stack: >
                at Object.eval (acceptance/chat-quoting-test:117:14)
        message: >
            navigates to the chatable url
        negative: >
            false
        browser log: |
    ...
error Command failed with exit code 1.
```

**Comment by martin-brennan** [comment]
Had a similar error to the quoting one locally, not sure what `global failure: Error` means:

```
not ok 67 Chrome 100.0 - [120049 ms] - Discourse Chat | Component | chat-user-avatar: user is online
    ---
        stack: >
                at http://localhost:7357/assets/chunk.vendors-node_modules_pretender_dist_pretender_es_js-node_modules_qunit_qunit_qunit_js-node_mo-a05e2f.453a7e9daf9c124ace08.js:5046:251
        message: >
            global failure: Error
        negative: >
            false
        browser log: |
    ...
```

The tests did not even finish running locally...

**Comment by martin-brennan** [comment]
Okay deleting `test/javascripts/components/chat-channel-preview-card-test.js`, the new test I added, locally makes the suite finish locally, but one of the quoting tests still fails:

```
not ok 264 Chrome 100.0 - [499 ms] - Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in
    ---
        actual: >
            /chat/channel/7/Uncategorized
        expected: >
            /c/uncategorized/1
        stack: >
                at Object.eval (acceptance/chat-quoting-test:117:14)
        message: >
            navigates to the chatable url
        negative: >
            false
        browser log: |
    ...
Testem finished with non-zero exit code. Tests failed.
```

@jjaffeux can you please look at the test I added in this PR and see if there is anything that jumps out as wrong with it? It seems so simple IDK why it would mess things up...

**Comment by jjaffeux** [comment]
@Chapoi maybe we limit padding on mobile screens, Im afraid it might take too much space on small screens?

File: assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  {{chat-channel-title channel=channel}}
+  {{#if hasDescription}}
+    <p class="chat-channel-preview-card__description">{{channel.description}}</p>
+  {{/if}}
+  {{d-button
+    action=(action "onJoinChannel")
+    label="chat.channel_settings.join_channel"
+    class="chat-channel-preview-card__join-channel-btn btn-primary"
+  }}
+  <p class="chat-channel-preview-card__browse-all">
```
Comment (Chapoi) [inline assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs:11]: Why is this wrapped in a p-tag? It's semantically not correct and it adds margin we don't want.
I would prefer the a-tag directly, without a wrapping element, but if for some ember-reason you need one, I'd prefer a simple span.

Comment (martin-brennan) [reply assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs:11]: No reason particularly, will change :)

Comment (martin-brennan) [reply assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs:11]: There is some margin/padding in the design that is lost if I remove the P tag though:

image

Here is what it looks like locally now:

image

If you want you can do another round of tweaks and push to this PR? I will just remove the P tag

**Comment by Chapoi** [comment]
> @Chapoi maybe we limit padding on mobile screens, Im afraid it might take too much space on small screens?

hm, it's looking fine to me like this
image

And since it's literally the main thing going on on the page, I think it's okay to take up this much space.

File: assets/javascripts/discourse/components/chat-channel-preview-card.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+      })
+      .catch(popupAjaxError);
+  }
+
+  @computed("channel.description")
+  get hasDescription() {
+    return !isEmpty(this.channel.description);
+  }
+
+  @computed("channel.description")
```
Comment (jjaffeux) [inline assets/javascripts/discourse/components/chat-channel-preview-card.js:39]: you can compute directly on hasDecription here

File: test/javascripts/acceptance/chat-test.js [hunk 10]
```diff
@@ -1334,15 +1326,6 @@ acceptance(
       });
     });
 
-    test("Composer placeholder is specific when previewing", async function (assert) {
```
Comment (jjaffeux) [inline test/javascripts/acceptance/chat-test.js:1337]: Should we remove chat.placeholder_previewing key now ?

File: test/javascripts/components/chat-channel-preview-card-test.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
+import { setupRenderingTest } from "discourse/tests/helpers/component-test";
```
Comment (jjaffeux) [inline test/javascripts/components/chat-channel-preview-card-test.js:1]: we should have a test for onJoinChannel

File: test/javascripts/acceptance/chat-browse-test.js [hunk 12]
```diff
@@ -77,6 +106,37 @@ acceptance("Discourse Chat - chat browsing", function (needs) {
       "Follow button is present"
     );
   });
+
+  test("Previewing a channel", async function (assert) {
+    await visit("/chat/browse");
+    await click(".chat-channel-unfollow");
+    await click(".chat-channel-preview");
+
+    assert.ok(
```
Comment (jjaffeux) [inline test/javascripts/acceptance/chat-browse-test.js:115]: I think we could get away with just this test here, we are actually kind duplicating tests here, we want to ensure the card is showing, we already have various tests to assert the card is acting as we want it to

File: assets/javascripts/discourse/templates/components/chat-live-pane.hbs [hunk 13]
```diff
@@ -1,5 +1,5 @@
 {{#if (and fullPage includeHeader)}}
-  <div class="chat-full-page-header {{if previewing "previewing"}}">
+  <div class="chat-full-page-header {{if previewing 'previewing'}}">
```
Comment (jjaffeux) [inline assets/javascripts/discourse/templates/components/chat-live-pane.hbs:2]: For now please fix it manually @martin-brennan, it's on @cvx plate to get linting on par between core and chat

**Comment by jjaffeux** [comment]
Tried on small screen and dark mode, looks great :heart:

Screenshot 2022-07-11 at 12 28 56

**Comment by jjaffeux** [comment]
ok tests should be fixed here, I don't have an exact answer on "why" for everything yet, I'm new to this new style too :p But I think it's good enough to move on.

@martin-brennan @Chapoi great work :+1:

**Review by jjaffeux — APPROVED** [review]
Just be sure to fix linting first

File: test/javascripts/components/chat-channel-preview-card-test.js (hunk 11 above)
Comment (martin-brennan) [reply test/javascripts/components/chat-channel-preview-card-test.js:1]: I am not sure the best way to test this for a component test? The ember guide mentions using service stubs which we do have for chat https://guides.emberjs.com/release/testing/testing-components/ but I'm not sure what exactly I should be testing wrt AJAX calls and things like that. The chat-browse-test already checks that the join button works and the card is gone / channel is reloaded. I am going to merge this PR, but would like your opinion on this and I can add another test afterward if needed.

Comment (jjaffeux) [reply test/javascripts/components/chat-channel-preview-card-test.js:1]: I think what we want to test here os that clicking the button calls an action named onJoinChannel

Comment (martin-brennan) [reply test/javascripts/components/chat-channel-preview-card-test.js:1]: There doesn't really seem to be an easy way to do that though? In https://guides.emberjs.com/release/testing/testing-components/#toc_testing-actions they presume that the action is being passed in via args like this:

```
export default class CommentFormComponent extends Component {
  @tracked comment = '';

  @action
  submitComment() {
    this.args.submitComment({ comment: this.comment });
  }
}
```

```
<CommentForm @submitComment={{this.externalAction}} />
```

Which is not the case for the preview card. I've tried a few different ways of overriding `onJoinChannel` in the test with a stub and none of them worked. And anyway not sure what value testing it here adds when we already test clicking the button in chat-browse-test? If you had an example of what you mean that would be great.

Comment (jjaffeux) [reply test/javascripts/components/chat-channel-preview-card-test.js:1]: I looked at this today and made it work but I want to write more dedicated guidelines for this. Im just like you and I'm learning to adapt to our new patterns and also improve what we currently do.

I think the good separation is:
- one happy path acceptance test
- lots of component/unit tests for various edge cases

I will elaborate more on this in the near future.

Comment (martin-brennan) [reply test/javascripts/components/chat-channel-preview-card-test.js:1]: Nice, thank you for looking into this!
